asmcomp/arm64/emit.mlp

```diff
@@ -318,6 +318,37 @@ let output_epilogue f =
   (* reset CFA back because function body may continue *)
   if n > 0 then cfi_adjust_cfa_offset n
 
+(* Output add-immediate / sub-immediate / cmp-immediate instructions *)
+
+let rec emit_addimm rd rs n =
+  if n < 0 then emit_subimm rd rs (-n)
+  else if n <= 0xFFF then
+    `	add	{emit_reg rd}, {emit_reg rs}, #{emit_int n}\n`
+  else begin
+    assert (n <= 0xFFF_FFF);
+    let nl = n land 0xFFF and nh = n land 0xFFF_000 in
+    `	add	{emit_reg rd}, {emit_reg rs}, #{emit_int nh}\n`;
+    if nl <> 0 then
+      `	add	{emit_reg rd}, {emit_reg rd}, #{emit_int nl}\n`
+  end
+
+and emit_subimm rd rs n =
+  if n < 0 then emit_addimm rd rs (-n)
+  else if n <= 0xFFF then
+    `	sub	{emit_reg rd}, {emit_reg rs}, #{emit_int n}\n`
+  else begin
+    assert (n <= 0xFFF_FFF);
+    let nl = n land 0xFFF and nh = n land 0xFFF_000 in
+    `	sub	{emit_reg rd}, {emit_reg rs}, #{emit_int nh}\n`;
+    if nl <> 0 then
+      `	sub	{emit_reg rd}, {emit_reg rd}, #{emit_int nl}\n`
+  end
```

Comment:
This code looks correct to me, but I'd feel better if there was a test for the new immediate codegen. (It wouldn't surprise me if there was nowhere in the testsuite that actually adds a constant above 4096 which is not divisible by 4096, in which case nothing in precheck would hit this path).

Reply:
I'll think about a more general test for immediate operands, not restricted to add/sub for ARM64.

Reply:
I've pushed a general test for immediate arguments.  Will check the "precheck" results tomorrow.